src/event_parser.rs

```diff
@@ -381,404 +357,222 @@ mod tests {
+        let mut parser = EventParser::new();
+        assert!(parser.process_bytes(Bytes::from("data:hello\n")).is_ok());
+        assert!(parser.process_bytes(Bytes::from("data:world\n\n")).is_ok());
         assert_eq!(
-            decoded.poll_next_unpin(&mut cx),
-            Ready(Some(Ok(event("message", &b"hello\nworld"[..], &b""[..]))))
+            parser.get_event(),
+            Some(event("message", &b"hello\nworld"[..], &b""[..]))
         );
-        assert_eq!(decoded.poll_next_unpin(&mut cx), Ready(None));
+        assert!(parser.get_event().is_none());
     }
 
-    #[test_case(b"\n\n\n\n" ; "all LFs")]
-    #[test_case(b"\r\r\r\r" ; "all CRs")]
-    #[test_case(b"\r\n\r\n\r\n\r\n" ; "all CRLFs")]
-    fn test_decode_repeated_terminators(bytes: &[u8]) {
-        use Poll::Ready;
-        let waker = noop_waker();
-        let mut cx = Context::from_waker(&waker);
+    #[test_case("\n\n\n\n" ; "all LFs")]
+    #[test_case("\r\r\r\r" ; "all CRs")]
+    #[test_case("\r\n\r\n\r\n\r\n" ; "all CRLFs")]
+    fn test_decode_repeated_terminators(chunk: &'static str) {
+        let mut parser = EventParser::new();
+        assert!(parser.process_bytes(Bytes::from(chunk)).is_ok());
 
         // spec seems unclear on whether this should actually dispatch empty events, but that seems
         // unhelpful for all practical purposes
-        let repeated_newlines = one_chunk(bytes);
-        assert_eq!(
-            Decoded::new(repeated_newlines).poll_next_unpin(&mut cx),
-            Ready(None)
-        );
+        assert!(parser.get_event().is_none());
     }
 
     #[test]
     fn test_decode_extra_terminators_between_events() {
-        use Poll::Ready;
-        let waker = noop_waker();
-        let mut cx = Context::from_waker(&waker);
-
-        let bytes = b"data: abc\n\n\ndata: def\n\n";
-
-        let mut decoded = Decoded::new(one_chunk(bytes));
-        let expected1 = event("message", &b"abc"[..], &b""[..]);
-        let expected2 = event("message", &b"def"[..], &b""[..]);
-        assert_eq!(decoded.poll_next_unpin(&mut cx), Ready(Some(Ok(expected1))));
-        assert_eq!(decoded.poll_next_unpin(&mut cx), Ready(Some(Ok(expected2))));
-        assert_eq!(decoded.poll_next_unpin(&mut cx), Ready(None));
-    }
-
-    #[test]
-    fn test_decode_edge_cases() {
-        use Poll::Ready;
-        let waker = noop_waker();
-        let mut cx = Context::from_waker(&waker);
-
-        let empty = stream::empty::<Result<Bytes>>();
-        assert_eq!(Decoded::new(empty).poll_next_unpin(&mut cx), Ready(None));
-
-        let one_empty = one_chunk(b"");
-        assert_eq!(
-            Decoded::new(one_empty).poll_next_unpin(&mut cx),
-            Ready(None)
-        );
+        let mut parser = EventParser::new();
+        assert!(parser
+            .process_bytes(Bytes::from("data: abc\n\n\ndata: def\n\n"))
+            .is_ok());
 
-        let one_comment_unterminated = one_chunk(b":hello");
```

Comment:
This is the one part that has me concerned. We are no longer testing when the stream gets interrupted because there is no stream. The ReconnectingRequest should probably have some unit or integration testing added to it since it currently has none.

I started exploring this but it seems like it's going to take some doing, and this PR is already getting large.